### PR TITLE
ts2.1: rearrange type translation, use new ObjectFlags

### PR DESCRIPTION
**Pull request by evmar**
Make the main "translate" function match the enum in the
TypeScript API.  Split out the handling of more complex
Object types into a separate function, and adjust the code
there to use the new objectFlags field.

This actually makes most tests pass(!) but there are still
a few more to track down, and there are still a bunch of
TODOs.

More work on #295.

### CODE REVIEW

**Comment by evmar** [comment]
Note: it's still intentional that the tests don't pass.

File: src/type-translator.ts [hunk 1]
```diff
@@ -164,51 +185,78 @@ export class TypeTranslator {
...
-      case ts.TypeFlags.Enum:
-        return 'number';
-      case ts.TypeFlags.StringLiteral:
-        return 'string';
-      case ts.TypeFlags.BooleanLiteral:
-        return 'boolean';
-      case ts.TypeFlags.NumberLiteral:
-        return 'number';
+      case ts.TypeFlags.Never:
+        this.warn(`should not emit a 'never' type`);
```
Comment (rkirov) [inline src/type-translator.ts]: why warning here? I think there are valid programs with the never type. Like users that mistrust the type system and want to augment it with a runtime check

```
switch (x)
 default:
  throw Error(x as any) <- x could be never here 
```

File: src/type-translator.ts [hunk 2]
```diff
@@ -230,34 +278,27 @@ export class TypeTranslator {
...
       // A reference to another type, e.g. Array<number> refers to Array.
       // Emit the referenced type and any type arguments.
       let referenceType = type as ts.TypeReference;
 
       let typeStr = '';
-      // TODO(ts2.1): let isTuple = (referenceType.flags & ts.TypeFlags.Tuple) > 0;
-      let isTuple = true;
-      // For unknown reasons, tuple types can be reference types containing a
-      // reference loop. see Destructuring3 in functions.ts.
-      // TODO(rado): handle tuples in their own branch.
```
Comment (rkirov) [inline src/type-translator.ts]: thanks for taking care of my todo :)

File: src/type-translator.ts (hunk 1 above)
Comment (evmar) [inline src/type-translator.ts]: The question is what we want to do when we hit one of these.  E.g. if you accidentally make an array of never, what type do we emit for Closure?

These warnings are never shown to a user and only show up inside the test suite, so it's probably harmless to leave it in... (?)

File: src/type-translator.ts [hunk 3]
```diff
@@ -164,51 +185,78 @@ export class TypeTranslator {
...
+      case ts.TypeFlags.NumberLiteral:
         return 'number';
       case ts.TypeFlags.Boolean:
+      case ts.TypeFlags.BooleanLiteral:
         return 'boolean';
+      case ts.TypeFlags.Enum:
+      case ts.TypeFlags.EnumLiteral:
+        return 'number';
+      case ts.TypeFlags.ESSymbol:
+        // NOTE: currently this is just a typedef for {?}, shrug.
```
Comment (evmar) [inline src/type-translator.ts]: Done.

File: src/type-translator.ts (hunk 2 above)
Comment (evmar) [inline src/type-translator.ts]: It's not done yet!  There's an empty spot where tuples need to be handled.  :P

File: src/type-translator.ts (hunk 1 above)
Comment (rkirov) [inline src/type-translator.ts]: So it is for inexpressible concepts in Closure. Makes sense, maybe StringLiterals and etc *Literals should also get a warning then? No need to take care in the cl.

**Pull request merged**